### DISCUSSION TITLE
AR mode

### DISCUSSION THREAD

**Discussion by RjvPivs** [korlibs/korge#2139] [Q&A] [answered]
Hello there! I`ve just started learning the engine and I want to ask if it is possible to create an AR mode with it.

**Comment by soywiz** [comment]
@RjvPivs by AR do you mean Augmented Reality?

**Reply by RjvPivs** [reply]
Yep

**Reply by soywiz** [reply]
Wow, that was quick.

AR is not supported right now. What would be your use-case?

**Reply by RjvPivs** [reply]
I want just to use camera as a background. Is this fully impossible?

**Reply by soywiz** [reply, marked as answer]
Oh I see. No, I didn't meant it was not possible, just that there is nothing specific implemented in KorGE about that.
You can use the camera API for the operating systems you are using and place them into a texture; for example Android and iOS and grab the image sequences and use them in code.
For example: https://developer.android.com/media/camera/camera-deprecated/camera-api

To avoid fetching pixels from the camera, if there is an API to get an OpenGL texture from the camera:
https://stackoverflow.com/questions/57331861/get-the-opengl-texture-handle-from-a-surfacetexture

Then you can use `ForcedTexNativeImage`:
https://github.com/korlibs/korge/blob/11993ae7828c96ee261b5d610904c8ca5430b8d2/korge/test/korlibs/korge/render/AgBitmapTextureManagerTest.kt#L85-L94

This class would give you a `Bitmap` instance whose pixels when rendering are provided by a specific OpenGL texture with a specific integral opengl id:

```kotlin
class FixedNativeImage(
    override val forcedTexId: Int = 100,
    override val forcedTexTarget: Int = KmlGl.TEXTURE_EXTERNAL_OES,
) : ForcedTexNativeImage(100, 100, false) {
    override val name: String get() = "FixedNativeImage"
}
```

And you should be able to use it places where you are using Bitmap, like in a `image`.

**Reply by RjvPivs** [reply]
Great, thank you! I`ll check and try this
